File: webhooks/deploy_wait_for_webhook_workflow.py
#!/usr/bin/env python3
"""Register and start a durable webhook-driven workflow.

The program keeps its local workers running until the workflow reaches the
WAIT_FOR_WEBHOOK task, then exits. Conductor keeps the workflow running and
resumes it after a callback is sent by ``send_webhook_payload.py``.
"""

import sqlite3
import time
from pathlib import Path
from typing import Any

from conductor.client.automator.task_handler import TaskHandler
from conductor.client.configuration.configuration import Configuration
from conductor.client.http.models import StartWorkflowRequest, Workflow
from conductor.client.orkes_clients import OrkesClients
from conductor.client.workflow.conductor_workflow import ConductorWorkflow
from conductor.client.workflow.task.timeout_policy import TimeoutPolicy
from conductor.client.workflow.task.wait_for_webhook_task import wait_for_webhook
from conductor.ai.agents import AgentRuntime

from settings import settings

from .utils.agent_task import AgentTask
from .utils.webhook_agent import webhook_agent
from .utils.workers import get_user_email, send_email

WORKFLOW_NAME = "wait_for_webhook_demo"
WORKFLOW_VERSION = 1
SEND_EMAIL_TASK_NAME = "send_email"
SEND_EMAIL_TASK_REF = "send_email_ref"
WAIT_TASK_REF = "wait_for_webhook_ref"

DATABASE_PATH = Path(__file__).resolve().parent / "utils" / "webhook_codelab_storage.db"
EMAIL_OUTPUT_FIELDS = ("sent_time", "subject", "recipients")

WORKFLOW_TIMEOUT_SECONDS = 7 * 24 * 60 * 60
READINESS_TIMEOUT_SECONDS = 60
POLL_INTERVAL_SECONDS = 1


def build_workflow(workflow_executor) -> ConductorWorkflow:
    """Build the workflow definition without registering or starting it."""
    workflow = ConductorWorkflow(
        name=WORKFLOW_NAME,
        version=WORKFLOW_VERSION,
        executor=workflow_executor,
    )
    workflow.description = "Durable wait-for-webhook example"
    workflow.timeout_seconds(WORKFLOW_TIMEOUT_SECONDS)
    workflow.timeout_policy(TimeoutPolicy.TIME_OUT_WORKFLOW)

    # Exercise 2: Iterate over all user_ids in settings.user_ids and get their email addresses.
    get_email_task = get_user_email(
        task_ref_name="get_user_email_ref",
        user_id=workflow.input("user_id"),
    )

    # Exercise 2: Use a DYNAMIC_FORK to send an email to each recipient.
    # Be sure to create a JoinTask to consolidate results after your DynamicForkTask.
    # Also ensure that your task_ref_name values are unique for each sent email, for example by
    # appending the user_id to the task_ref_name.
    send_email_task = send_email(
        task_ref_name=SEND_EMAIL_TASK_REF,
        recipients=get_email_task.output("result"),
        subject="Hello from Alex",
        body="foo bar qua",
    )

    webhook_wait = wait_for_webhook(
        task_ref_name=WAIT_TASK_REF,
        matches={
            "$['type']": "customer",
            # Exercise 2: Change to 'user_ids'
            # Be sure the payload sent by send_webhook_payload.py matches the key you use here.
            "$['user_id']": workflow.input("user_id"),
        },
    )

    agent_task = AgentTask(
        task_ref_name="process_webhook_ref",
        agent_name=webhook_agent.name,
        prompt=webhook_wait.output("agent_input"),
    )

    workflow >> get_email_task >> send_email_task >> webhook_wait >> agent_task
    workflow.output_parameter("agent_response", agent_task.output("text"))

    return workflow


def start_workflow(workflow_client) -> str:
    """Start one workflow execution and return its execution ID."""
    request = StartWorkflowRequest(input={"user_id": settings.user_id})
    request.name = WORKFLOW_NAME
    request.version = WORKFLOW_VERSION

    return workflow_client.start_workflow(start_workflow_request=request)


def wait_until_webhook_ready(workflow_client, workflow_id: str) -> Workflow:
    """Return the execution after it reaches its WAIT_FOR_WEBHOOK task."""
    deadline = time.monotonic() + READINESS_TIMEOUT_SECONDS

    while time.monotonic() < deadline:
        execution = workflow_client.get_workflow(
            workflow_id,
            include_tasks=True,
        )

        wait_task = next(
            (
                task
                for task in execution.tasks or []
                if task.reference_task_name == WAIT_TASK_REF
            ),
            None,
        )

        if wait_task and wait_task.status == "IN_PROGRESS":
            print(f"{WAIT_TASK_REF} is ready")
            print(f"Webhook URL: {settings.webhook_endpoint_url}")
            return execution

        if execution.status in {"FAILED", "TIMED_OUT", "TERMINATED"}:
            raise RuntimeError(f"Workflow entered terminal status {execution.status}")

        time.sleep(POLL_INTERVAL_SECONDS)

    raise TimeoutError(
        f"Workflow did not reach {WAIT_TASK_REF} within "
        f"{READINESS_TIMEOUT_SECONDS} seconds"
    )


def get_completed_email_outputs(
    execution: Workflow,
) -> list[dict[str, Any]]:
    """Return valid outputs from every completed send-email task."""
    email_tasks = [
        task
        for task in execution.tasks or []
        if task.task_def_name == SEND_EMAIL_TASK_NAME and task.status == "COMPLETED"
    ]
    if not email_tasks:
        raise RuntimeError(
            f"No completed {SEND_EMAIL_TASK_NAME} task outputs were found"
        )

    email_outputs = []
    for task in email_tasks:
        output = task.output_data
        if not isinstance(output, dict):
            raise RuntimeError(
                f"Completed task {task.reference_task_name} has invalid output"
            )

        missing_fields = set(EMAIL_OUTPUT_FIELDS) - output.keys()
        if missing_fields:
            fields = ", ".join(sorted(missing_fields))
            raise RuntimeError(
                f"Completed task {task.reference_task_name} is missing: {fields}"
            )

        email_outputs.append(output)

    return email_outputs


def store_email_outputs(email_outputs: list[dict[str, Any]]) -> int:
    """Insert all completed email outputs in one database transaction."""
    if not email_outputs:
        raise ValueError("At least one email output is required")

    if not DATABASE_PATH.is_file():
        raise FileNotFoundError(
            f"Database not found at {DATABASE_PATH}. "
            "Run webhooks/utils/create_sqlite_db.py first."
        )

    email_rows = [
        tuple(email_output[field] for field in EMAIL_OUTPUT_FIELDS)
        for email_output in email_outputs
    ]

    with sqlite3.connect(DATABASE_PATH) as connection:
        connection.executemany(
            """
            INSERT INTO emails (
                sent_time,
                subject,
                recipients
            )
            VALUES (?, ?, ?)
            """,
            email_rows,
        )

    return len(email_rows)


def main() -> None:
    config = Configuration()
    task_handler = TaskHandler(
        workers=[],
        configuration=config,
        scan_for_annotated_workers=True,
    )
    task_handler.start_processes()

    try:
        clients = OrkesClients(configuration=config)
        # Deploy agent definition
        with AgentRuntime(configuration=config) as runtime:
            runtime.deploy(webhook_agent)
        workflow = build_workflow(clients.get_workflow_executor())
        workflow.register(overwrite=True)
        print(f"Registered workflow {WORKFLOW_NAME}, version {WORKFLOW_VERSION}")

        workflow_client = clients.get_workflow_client()
        workflow_id = start_workflow(workflow_client)

        workflow_url = f"{config.ui_host.rstrip('/')}/execution/{workflow_id}"
        print(f"Workflow URL: {workflow_url}")

        execution = wait_until_webhook_ready(workflow_client, workflow_id)
        email_outputs = get_completed_email_outputs(execution)
        stored_count = store_email_outputs(email_outputs)
        print(f"Stored {stored_count} email record(s) in {DATABASE_PATH}")
    finally:
        task_handler.stop_processes()


if __name__ == "__main__":
    main()

File: webhooks/java/src/main/java/io/orkes/codelab/webhooks/DeployWaitForWebhookWorkflow.java
package io.orkes.codelab.webhooks;

import com.netflix.conductor.client.http.WorkflowClient;
import com.netflix.conductor.common.metadata.tasks.Task;
import com.netflix.conductor.common.metadata.workflow.WorkflowDef;
import com.netflix.conductor.common.run.Workflow;
import com.netflix.conductor.sdk.workflow.def.ConductorWorkflow;
import com.netflix.conductor.sdk.workflow.def.tasks.SimpleTask;
import com.netflix.conductor.sdk.workflow.executor.WorkflowExecutor;

import io.orkes.codelab.webhooks.utils.AgentTask;
import io.orkes.codelab.webhooks.utils.WaitForWebhookTask;
import io.orkes.codelab.webhooks.utils.WebhookAgent;
import io.orkes.codelab.webhooks.utils.Workers;

import org.conductoross.conductor.ai.AgentRuntime;

import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeoutException;

/**
 * Register and start a durable webhook-driven workflow, running its workers until it reaches the
 * WAIT_FOR_WEBHOOK task. Conductor then keeps it suspended until the send-webhook step resumes it.
 */
public final class DeployWaitForWebhookWorkflow {

    private static final String WORKFLOW_NAME = "wait_for_webhook_demo_" + Settings.LANGUAGE;
    private static final int WORKFLOW_VERSION = 1;
    private static final String WAIT_TASK_REF = "wait_for_webhook_ref";

    private static final long WORKFLOW_TIMEOUT_SECONDS = Duration.ofDays(7).toSeconds();
    private static final Duration READINESS_TIMEOUT = Duration.ofSeconds(60);
    private static final Duration READINESS_POLL_INTERVAL = Duration.ofSeconds(1);
    private static final int WORKER_POLL_INTERVAL_MILLIS = 100;

    // Shown in the workflow's description in the Orkes Conductor UI, and in every
    // email this version sends, so you can tell the language versions apart.
    private static final String CODELAB_LANGUAGE = "Java";

    private DeployWaitForWebhookWorkflow() {}

    /** Build the workflow definition without registering or starting it. */
    private static ConductorWorkflow<Map<String, Object>> buildWorkflow(
            WorkflowExecutor workflowExecutor) {
        var workflow = new ConductorWorkflow<Map<String, Object>>(workflowExecutor);
        workflow.setName(WORKFLOW_NAME);
        workflow.setVersion(WORKFLOW_VERSION);
        workflow.setDescription(
                "Durable wait-for-webhook example (registered from " + CODELAB_LANGUAGE + ")");
        // Mark the execution TIMED_OUT if it runs longer than WORKFLOW_TIMEOUT_SECONDS,
        // for example because no webhook arrives. ALERT_ONLY would let it keep running.
        workflow.setTimeoutPolicy(WorkflowDef.TimeoutPolicy.TIME_OUT_WF);
        workflow.setTimeoutSeconds(WORKFLOW_TIMEOUT_SECONDS);

        String subject = "Hello from " + CODELAB_LANGUAGE;
        String body = "Sent by the " + CODELAB_LANGUAGE + " version of the webhooks codelab.";

        // Exercise 2: Replace this with one get_user_emails task that resolves every address in
        // ConductorWorkflow.input.get("user_ids"), so the number of emails is decided at runtime.
        // Pass it the subject and body as well, for the forked send_email tasks.
        var getEmailTask =
                new SimpleTask("get_user_email", "get_user_email_ref")
                        .input("user_id", ConductorWorkflow.input.get("user_id"));

        // Exercise 2: Send the emails with new DynamicFork(referenceName, getEmailsTask). It adds
        // get_user_emails before itself and a join after, so add only the fork to the workflow.
        var sendEmailTask =
                new SimpleTask("send_email", "send_email_ref")
                        .input("recipients", getEmailTask.taskOutput.get("result"))
                        .input("subject", subject)
                        .input("body", body);

        Map<String, Object> matches = new LinkedHashMap<>();
        // Every language version shares one webhook, so only match
        // payloads sent by this language's send-webhook step.
        matches.put("$['language']", Settings.LANGUAGE);
        matches.put("$['type']", "customer");
        // Exercise 2: Change to 'user_ids'. Be sure the payload sent by
        // SendWebhookPayload.java matches the key you use here.
        matches.put("$['user_id']", ConductorWorkflow.input.get("user_id"));
        var webhookWait = new WaitForWebhookTask(WAIT_TASK_REF, matches);

        var agentTask =
                new AgentTask(
                        "process_webhook_ref",
                        WebhookAgent.NAME,
                        webhookWait.taskOutput.get("agent_input"));

        workflow.add(getEmailTask);
        workflow.add(sendEmailTask);
        workflow.add(webhookWait);
        workflow.add(agentTask);
        workflow.setWorkflowOutput(Map.of("agent_response", agentTask.taskOutput.get("text")));

        return workflow;
    }

    /** Start one workflow execution and return its execution ID. */
    private static String startWorkflow(WorkflowExecutor workflowExecutor) {
        return workflowExecutor.startWorkflow(
                WORKFLOW_NAME, WORKFLOW_VERSION, Map.of("user_id", Settings.USER_ID));
    }

    /** Wait until the execution reaches its WAIT_FOR_WEBHOOK task. */
    private static void waitUntilWebhookReady(WorkflowClient workflowClient, String workflowId)
            throws InterruptedException, TimeoutException {
        // Unlike Instant.now(), System.nanoTime() is unaffected by changes to the system clock.
        long deadline = System.nanoTime() + READINESS_TIMEOUT.toNanos();

        while (System.nanoTime() - deadline < 0) {
            Workflow execution = workflowClient.getWorkflow(workflowId, true);

            boolean waiting =
                    execution.getTasks().stream()
                            .anyMatch(
                                    task ->
                                            WAIT_TASK_REF.equals(task.getReferenceTaskName())
                                                    && task.getStatus() == Task.Status.IN_PROGRESS);
            if (waiting) {
                // Exercise 1: Return this execution, with its tasks, so the caller can read every
                // send_email output (change this method to return Workflow).
                return;
            }

            if (execution.getStatus().isTerminal()) {
                throw new IllegalStateException(
                        "Workflow entered terminal status " + execution.getStatus());
            }

            Thread.sleep(READINESS_POLL_INTERVAL);
        }

        throw new TimeoutException(
                "Workflow did not reach %s within %d seconds"
                        .formatted(WAIT_TASK_REF, READINESS_TIMEOUT.toSeconds()));
    }

    public static int run() throws InterruptedException, TimeoutException {
        Settings settings = Settings.current();
        // Build the agent first, so a malformed model setting fails before any worker starts.
        var agent = WebhookAgent.create();
        var apiClient = settings.createApiClient();
        var workflowExecutor = new WorkflowExecutor(apiClient, WORKER_POLL_INTERVAL_MILLIS);

        try {
            // Starts polling for the tasks of every @WorkerTask method in Workers.
            workflowExecutor.initWorkersFromInstances(List.of(new Workers()));

            // Register the agent that the workflow's AGENT task runs. The runtime gets a client
            // of its own, because closing the runtime also shuts down the client it was given.
            try (var runtime = new AgentRuntime(settings.createApiClient())) {
                runtime.deploy(agent);
            }

            // Overwrite any earlier version, and register a task definition for each SIMPLE task
            // that the server does not know yet.
            buildWorkflow(workflowExecutor).registerWorkflow(true, true);
            System.out.println(
                    "Registered workflow " + WORKFLOW_NAME + ", version " + WORKFLOW_VERSION);

            String workflowId = startWorkflow(workflowExecutor);
            System.out.println(
                    "Workflow URL: " + settings.serverBaseUrl() + "/execution/" + workflowId);

            waitUntilWebhookReady(new WorkflowClient(apiClient), workflowId);
            // Exercise 1: Store a row in QuerySqliteDb.DATABASE_PATH for each completed send_email
            // task, matching on getTaskDefName(): each task's getReferenceTaskName() is unique from
            // Exercise 2 on. Read each result from getOutputData(); numbers are Integer or Long.
            System.out.println(WAIT_TASK_REF + " is ready");
            System.out.println(
                    "Before sending the webhook, start the agent tool workers with "
                            + "`./gradlew serve-agent`.");
            System.out.println("Webhook URL: " + settings.webhookEndpointUrl());
            return 0;
        } finally {
            workflowExecutor.shutdown();
            apiClient.shutdown();
        }
    }
}
